File: android/src/main/java/com/jvanila/flutter/shopify/usecases/CreateCustomerAddressUseCase.java
package com.jvanila.flutter.shopify.usecases;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.jvanila.flutter.plugin.arch.PluginContext;
import com.shopapp.gateway.ApiCallback;
import com.shopapp.gateway.entity.Address;
import com.shopapp.gateway.entity.Error;
import com.shopapp.shopify.api.ShopifyApi;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class CreateCustomerAddressUseCase extends ShopifyCallUseCase {

    private static final String ARG_ADDRESS_JSON = "addressJson";

    public CreateCustomerAddressUseCase(PluginContext<ShopifyApi> pluginContext) {
        super(pluginContext);
    }

    @Override
    protected void call(MethodCall input, final MethodChannel.Result result) {

        String addressJson = input.argument(ARG_ADDRESS_JSON);

        Address addr = new Gson().fromJson(addressJson,
                new TypeToken<Address>() {
                }.getType());

        mPluginContext.api.instance.createCustomerAddress(addr, new ApiCallback<String>() {
            @Override
            public void onResult(String addressId) {
                result.success(addressId);
            }

            @Override
            public void onFailure(Error error) {
                System.out.println("onFailure -- " + error);
                result.error("CreateCustomerAddressUseCase", error.getMessage(), error);
            }
        });
    }
}

File: ios/Classes/UseCases/CreateCustomerAddressUseCase.swift
//
//  CreateCustomerAddressUseCase.swift
//  shopify
//
//  Created by Joseph jose on 10/09/19.
//

import UIKit
import ShopApp_Shopify
import ShopApp_Gateway
import Flutter

class CreateCustomerAddressUseCase: UseCase {

    static let ARG_ADDRESS_JSON = "addressJson"
    
    override public init(_ context: PluginContext) {
        super.init(context)
    }
    
    override public func trigger(with methodCall: FlutterMethodCall, result: @escaping (Any?) -> Void) {
        
        if let args = methodCall.arguments as? [String:String] {
            if let addressJson = args[CreateCustomerAddressUseCase.ARG_ADDRESS_JSON] {
                
                let addr = Address(addressJSON: addressJson)
                (mContext.api.instance as! ShopifyAPI).addCustomerAddress(with: addr) { (addressId, error) in
                    
                    if error == nil && addressId != nil {
                        
                        result(addressId)
                    }else if addressId == nil {
                        
                        result(false)
                    }else {
                        
                        self.createError(withCase: "createCustomerAddressUseCase", message: error!.errorMessage, error: error!, on: result)
                    }
                }
            }
        }
    }
}
